### ISSUE TITLE
Division by zero if o.W != 0 results in `NaN` values

### ISSUE THREAD

**Issue by rlefmann** [gdkrmr/WeightedOnlineStats.jl#42]
Some of the algorithms, like `WeightedMean` and `WeightedCovMatrix`, return `NaN` values if the first weight is zero. Example:

```julia
using WeightedOnlineStats

m = WeightedMean()
x = ones(3)
w = [0.0, 0.5, 1.0]
fit!(m, x, w)
# WeightedMean: ∑wᵢ=1.5 | value=NaN

# but if we reverse the sequences everything works as expected:
m = WeightedMean()
fit!(m, reverse(x), reverse(w))
# WeightedMean: ∑wᵢ=1.5 | value=1.0
```

If at any point in the computation the sum of the previous weights, `o.W`, becomes zero, there will be `NaN` values because of division by zero. For the weighted mean this could be fixed by not changing `mu` in that case.

**Comment by gdkrmr** [comment]
Good catch. Thanks for reporting. The whole thing was written without zero weights in mind. 

Now there is the non-obvious questions: Should we increase the counter of observations when encountering a zero weight or just not do anything at all? 

We also don't guard for negative weights, so should we guard for zero weights? I personally lean in favor of leaving this in the responsibility of the user.

**Comment by tpapp** [comment]
Friendly bump, I am running into this occasionally.

Looking at the implementation, I am wondering if keeping the *average* weight makes sense. Might as well record the *total* weight.

FWIW, I would bump `n` every time `fit!` is called.

**Comment by gdkrmr** [comment]
Easy things first: `o.W` is the average weight, you can calculate the total weight by doing `o.W * o.n`, there is even an exported function, `weightsum` if you require to do this.

Keeping track of the average instead of the total weight is necessary because recording `o.W` as the total sum of weights makes `o.W` growing large compared to `w[i]` which causes catastrophic cancellation. This was the whole reason for me to write this package :-). Else you could just track `S += w[i] * x[i]` and `W += w[i]` and do `S/W` in the end and be done with it, which is much more intuitive from a user standpoint. The assumption here is that all `w[i]` are more or less of the same order of magnitude.

Much more complicated is the zero weight issue and it is not quite clear to me. Therefore I think leaving the interpretation up to the user and just focus on a numerically accurate and performant implementation is not the worst of choices.

I am trying to wrap my head around it, please chime in @tpapp and everyone who works with values that have weight zero but I am not sure if the added complexity and reduced flexibility for the user is worth it:

- what does it mean semantically if I have a value with weight 0? 
- Why would we even want to count, i.e. increase `n`, when we encounter a value with weight zero?
- Is there any ambiguity? I.e. could there be a case where one person wants to count the zero weight value and another one doesn't want to count it.
- especially: what is the expected value of `mean(o)` here?

```julia
o = WeightedMean()
fit!(o, 1, 0)
mean(o) # what is the expected value here?
```

From a numerical perspective: `0/0 === NaN` and `NaN/0 === NaN` and therefore `mean(o) === NaN` is the correct value. The issue arises because `WeightedMean()` cannot recover once `o.μ === NaN`. Getting this right also means that we may reconsider how we initialize `o.µ` and `o.W`, it may make sense to initialize both as `NaN`.

An alternative interpretation could be that without any information we expect a mean of zero and that fitting values is like updating our hypothesis on the expected value. Bayesian statistics always has been my weak point and I don't know if this changes anything to the numerics.

Final issue: the library is designed to work with all sorts of number types, I have personally used it with `DoubleFloats.jl` and `BigFloat` but there are Number types that don't have NaNs, e.g. fixed point numbers and posits, I haven't used it with any of those but I also don't want to prevent anyone from using those number types.

**Comment by tpapp** [comment]
Sorry, I missed this reply somehow, and just saw #62.

My preferred solution would be to simply *skip* updating when the weight is zero, with an `if` clause. That solves all complications for me.

As for the interpretation of zero weights: I am doing calculations where some analytical probabilities may be zero, and the motivation is to keep using the same code. So, as far as I am concerned, a zero weight should do nothing.

FWIW, I think the count is not well-defined with weighted statistics, so I am not using it.

**Comment by gdkrmr** [comment]
I think #64 makes me happy. @tpapp does that work for you? The rationale is that a zero weights does not change the mean, o.W and o.n still get updated. I also found some minor bugs and improved testing.

**Comment by tpapp** [comment]
Broadly yes, but I did not work though the math. I commented in the PR, I am not sure I understand all the invariants.

**Comment by gdkrmr** [comment]
I do quite some numerical testing, also with low accuracy numbers, so would say that this is ready to merge. 

- I have opened #66 for better numerical testing
- There is an issue on better documentation #46 

so, if there are no problems with this until early next week, I'll merge #64
